doc/theory_background.md

```diff
@@ -0,0 +1,55 @@
+# iDynFor theory background
+
+This document provide an overview of the different semantics between iDynTree and Pinocchio
+data-structures and interfaces, and provide a reference for the implentation in iDynFor classes, that implements iDynTree-inspired interfaces using Pinocchio.
+
+## iDynTree::Model vs pinocchio::Model
+
+### Link and Joints
+
+In `iDynTree::Model` a multibody is composed by its links that are interconnected by joints.
+Each link is connected to another link by a joint, so you can't attach a joint to another joint
+without having a link in the middle. Furthermore, `iDynTree::Model` consider all models to be
+floating base by default, so there is no "joint" connecting the base to "universe" or "world" frame. Only links have a frame w.r.t. to which you can compute forward kinematics or the jacobian, while there is no explicit concept of "joint frame"
+
+On the other hand, in `pinocchio::Model` joints are interconnected to each other, to build
+the so-called "kinematic tree". So multiple joints can be connected to each other, without having a body in the middle. On the other hand, each link must have a parent joint, even if it is the first body of the kinematic tree.
+A model can be connected to the "universe" reference frame with any kind of frame, but for consistency with `iDynTree` the pinocchio models built in `iDynFor` are always connected to "universe" with a `pinocchio::JointFreeFlyer` (i.e. 6-DOF joint).
+
+For these reasons, the count of links and joints in `iDynTree::Model` and  `pinocchio::Model` are differents.
+For links, in iDynTree only the internal bodies that compose the multibody model are counted, while for `pinocchio` also the `universe` "link" is considered.
+For joints, in iDynTree only the joints that internconnect internal links are considered, while in `pinocchio` two additional joints are considerd:
+* the joint that connects the floating base to `universe`,
+* the  parent joint of `universe`.
+
+We will always have that for corresponding models:
+* `iDynTree::Model::getNrOfLinks() + 1 == pinocchio::Model::nbodies` and
+* `iDynTree::Model::getNrOfJoints() + 2 == pinocchio::Model::njoints`.
```

Comment:
Why ``+2``? Does ``pinocchio`` add a 3D prismatic joint and a spherical joint? This does not seem to match with the sentence

> but for consistency with `iDynTree` the pinocchio models built in `iDynFor` are always connected to "universe" with a `pinocchio::JointFreeFlyer` (i.e. 6-DOF joint).

Reply:
That was confusing also for me when I originally wrote the document. Basically, the key is that in pinocchio "each link must have a parent joint, even if it is the first body of the kinematic tree." So if one has a model with just a single link, that has zero joints in iDynTree, in pinocchio it contains two joints:
* the joint that connects the floating base to `universe`,
* the  parent joint of `universe`.

I tried to write this in the document as:

> For joints, in iDynTree only the joints that internconnect internal links are considered, while in `pinocchio` two additional joints are considerd:
> * the joint that connects the floating base to `universe`,
> * the  parent joint of `universe`.

but probably it is betterto rephrase that part?

Reply:
I tried to clarify the concepts in https://github.com/ami-iit/idynfor/pull/21/commits/a1f017c76d32d56cd9c1b9ad72273061a14612f3 .

Reply:
Oh, that's confusing. Can the parent joint of ``universe`` move then? :thinking:

Reply:
Now, it is forced to be fixed in the constructor: https://github.com/stack-of-tasks/pinocchio/blob/v2.6.17/src/multibody/model.hpp#L186 .

(By looking for this, I also found https://github.com/stack-of-tasks/pinocchio/blob/78d62096002ffa3790638e392f0b6e4a5efc3d34/doc/Overview.md?plain=1#L143 that is related).

Reply:
At least all the confusion is motivating the reason behind iDynFor. : )